### PR TITLE
Fix runtime cyclic readonly tuple type def failure

### LINKED ISSUE #40664: [Bug]: Runtime error for cyclic readonly tuple types.
### Description

type A1 [A1] & readonly;

For the above case  we are getting

```
[2023-06-12 17:40:15,975] SEVERE {b7a.log.crash} - $type$A1 
java.lang.NoSuchFieldError: $type$A1
        at prakanth.test.0.constants.$_typeref_type_constants.$populate$typeRefType$A1(Unknown Source)
        at prakanth.test.0.constants.$_typeref_type_constants.$populate_typeref_types(Unknown Source)
        at prakanth.test.0.types.$_types.$createTypeConstants(Unknown Source)
        at prakanth.test.0.types.$_types.$createTypes(Unknown Source)
        at prakanth.test.0.$_init.$currentModuleInit(test)
        at prakanth.test.0.$configurationMapper.$initAndPopulateConfigData(Unknown Source)
        at prakanth.test.0.$configurationMapper.$configureInit(Unknown Source)
        at prakanth.test.0.$_init.main(test)
```

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Fixes #40664 

## Approach
Created type definition for effective type of the immutable tuple types.

## Samples
```
type A [A] & readonly;
```
Previously it results in runtime error. This will work with this fix.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/util/Flags.java [hunk 1]
```diff
@@ -81,6 +81,7 @@ public class Flags {
     public static final long INFER = ANY_FUNCTION << 1;                         //  40
     public static final long ENUM_MEMBER = INFER << 1;                          //  41
     public static final long QUERY_LAMBDA = ENUM_MEMBER << 1;                   //  42
+    public static final long TYPE_DEF = QUERY_LAMBDA << 1;                         //  43
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/util/Flags.java:84]:
```suggestion
    public static final long TYPE_DEF_CREATED= QUERY_LAMBDA << 1;                         //  43
```

May be?

Comment (ushirask) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/util/Flags.java:84]: +1, would be more meaningful.

Comment (hasithaa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/util/Flags.java:84]: +1

File: tests/jballerina-unit-test/src/test/resources/test-src/typedefs/cyclic-type-definitions.bal [hunk 2]
```diff
@@ -210,6 +210,46 @@ function testCyclicObjectTypeDefinition() {
...
+type T2 T3 & readonly;
+type T3 [T2?] & readonly;
+type T4 [T4?, 1, "a"] & readonly;
+
+type T5 [(T5 & readonly)?];
+
+public function testCyclicReadonlyTupleTypeDefinition1() {
+    T1 t1 = [];
+    assertTrue(t1 is readonly);
+    assertEquals(t1.toString(), "[null]");
```
[the remark this answers is not in the document]
Comment (hasithaa) [reply tests/jballerina-unit-test/src/test/resources/test-src/typedefs/cyclic-type-definitions.bal:224]: Seem it uses an informal style. This is ok.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/util/Flags.java (hunk 1 above)
Comment (prakanth97) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/util/Flags.java]: Updated [11da802](https://github.com/ballerina-platform/ballerina-lang/pull/40678/commits/11da8025f891454affd08a73696b7a0b8babb281)

**Review by hasithaa — APPROVED** [review]

**Pull request merged**